### PR TITLE
got all Street Address Match Tests to Pass

### PR DESCRIPTION
**Pull request by vardahoth**

### CODE REVIEW

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java [hunk 1]
```diff
@@ -2,13 +2,185 @@
...
+        put("trail", "trail");
+        put("cir", "circle");
+        put("circle", "circle");
+        put("blvd", "boulevard");
+        put("boulevard", "boulevard");
+        put("rd", "road");
+        put("road", "road");
+    }};
+
+    private static final Hashtable<String,String> compassDirection = new Hashtable<String,String>() {{
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/StreetAddressMatchLogic.java]: FYI... Some cities also use NE, NW, SE, SW. I can't say that I've ever seen them spell out North-East in an address though.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java [hunk 2]
```diff
@@ -2,13 +2,185 @@
...
+        // get return on po box addressses (if either one is a po box)
+        if (isAddressPoBox(stAddressParts1) && isAddressPoBox(stAddressParts2)){
+            // Strip all characters out of address1, and only leave numbers
+            for (int part = 0; part < stAddressParts1.size(); part++){
+                if (!isInteger(stAddressParts1.get(part))){
+                    stAddressParts1.remove(part);
+                    part--;
+                }
+            }
+            // Strip all characters out of address2, and only leave numbers
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/StreetAddressMatchLogic.java]: Duplicate code... Should move into a separate helper function. Added benefit is that you can directly test the routine and make sure you hit all edge cases.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java [hunk 3]
```diff
@@ -2,13 +2,185 @@
...
+            if (getCompassDirection().containsKey(stPart)){
+                st1Direction = getCompassDirection().get(stPart);
+                stAddressParts1.remove(stDirectionElement); // remove the direction from the address
+                break;
+            }
+            stDirectionElement++;
+        }
+        stDirectionElement = 0;
+
+        for (String stPart: stAddressParts2){
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/StreetAddressMatchLogic.java]: Also duplicate code from above? Split into helper method.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java [hunk 4]
```diff
@@ -2,13 +2,185 @@
...
+                return AnswerType.no;
+            }
+        } else if (isAddressPoBox(stAddressParts1) ^ isAddressPoBox(stAddressParts2)){
+            return AnswerType.no;
+        }
+
+
+        // abstract street direction if the address has one
+        int stDirectionElement = 0;
+        for (String stPart: stAddressParts1){
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/StreetAddressMatchLogic.java]: Can't we just take the stAddressParts1 and remove all the CompassDirections? Something like

for (String roadType: roadTypes) { stAddressParts1.remove(roadType) }  

It would have to be rewritten to work with a HashTable or put the directions in a List.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java [hunk 5]
```diff
@@ -2,13 +2,185 @@
...
+    public static AnswerType getStreetAddressMatchResult(ArrayList<String> stAddressParts1, ArrayList<String> stAddressParts2){
+        // Precondition: All addressParts must follow the rules from the setNormalizeStreetAddress()
+        String st1Direction = "";
+        String st2Direction = "";
+        String st1RoadType = "";
+        String st2RoadType = "";
+
+        // get return on po box addressses (if either one is a po box)
+        if (isAddressPoBox(stAddressParts1) && isAddressPoBox(stAddressParts2)){
+            // Strip all characters out of address1, and only leave numbers
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/StreetAddressMatchLogic.java]: If your comment is accurate, you can do the same thing with the following code:

        StringBuilder sb = new StringBuilder();
        for (char c: buffer.toCharArray()) { if (Character.isDigit(c)) { sb.append(c); } }

There should be a way to turn this into a single line lambda expression but my first attempt failed at it is getting late.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java (hunk 1 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/StreetAddressMatchLogic.java]: hmmm, if you think I should include them in street address, I will. For now I'll leave them left out.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java (hunk 2 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/StreetAddressMatchLogic.java]: I thought about this... but it's doing two things...
1. seeing if the address has a direction, and storing that direction for comparison later....
2. removing the direction from the address array.

I couldn't think of a way two return both of these at the same time. The only option I could think of was having two separate methods for each of these... but that would mean 2 extra for loops (4 total). Maybe we can discuss this Monday.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java (hunk 3 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/StreetAddressMatchLogic.java]: This one may be a bit more tricky since it's doing two things. If I seperated this into functions, I would have to write two methods for each address (4 for loops) just for the direction.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java (hunk 4 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/StreetAddressMatchLogic.java]: I can do this... but currently the match is set to yes, if they are the same... and maybe if one is missing. So I'm not sure what the blanket answer should be.

File: src/com/cft/contactmerge/StreetAddressMatchLogic.java (hunk 5 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/StreetAddressMatchLogic.java]: I'll look into this.
